File: PyLibs/src/Libs/FrameHandle/TimerMgr.py
# -*- coding: utf-8 -*-

from Libs.DelayHandle.DelayHandleMgrBase import DelayHandleMgrBase;
from Libs.DataStruct.MList import MList;
from Libs.Tools.UtilApi import UtilApi;

'''
@brief 定时器管理器
'''

class TimerMgr(DelayHandleMgrBase):

    def __init__(self):
        super(TimerMgr, self).__init__();
        
        self.mTypeId = "TimerMgr";
        
        self.mTimerList = MList();     # 当前所有的定时器列表


    def addObject(self, delayObject, priority = 0.0):
        # 检查当前是否已经在队列中
        if (self.mTimerList.IndexOf(delayObject) == -1):
            if (self.bInDepth()):
                super(TimerMgr, self).addObject(delayObject, priority);
            else:
                self.mTimerList.Add(delayObject);


    def removeObject(self, delayObject):
        # 检查当前是否在队列中
        if (self.mTimerList.IndexOf(delayObject) != -1):
            delayObject.mIsDisposed = True;
            if (self.bInDepth()):
                super(TimerMgr, self).removeObject(delayObject);
            else:
                for item in self.mTimerList.getList():
                    if (UtilApi.isAddressEqual(item, delayObject)):
                        self.mTimerList.Remove(item);
                        break;


    # 从 Lua 中添加定时器，这种定时器尽量整个定时器周期只与 Lua 通信一次
    def addTimer(self, delayObject, priority = 0.0):
        self.addObject(delayObject, priority);


    def removeTimer(self, timer):
        self.removeObject(timer);


    def Advance(self, delta):
        self.incDepth();

        for timerItem in self.mTimerList.getList():
            if (not timerItem.IsClientDispose()):
                timerItem.OnTimer(delta);

            if (timerItem.mIsDisposed):        # 如果已经结束
                self.removeObject(timerItem);

        self.decDepth();

File: JSLib/src/FrameHandle/TimerMgr.js
using LuaInterface;
using System;

/**
 * @brief 定时器管理器
 */
namespace SDK.Lib
{
    public class TimerMgr : DelayHandleMgrBase
    {
        protected MList<TimerItemBase> m_timerList;     // 当前所有的定时器列表

        public TimerMgr()
        {
            m_timerList = new MList<TimerItemBase>();
        }

        protected override void addObject(IDelayHandleItem delayObject, float priority = 0.0f)
        {
            // 检查当前是否已经在队列中
            if (m_timerList.IndexOf(delayObject as TimerItemBase) == -1)
            {
                if (bInDepth())
                {
                    base.addObject(delayObject, priority);
                }
                else
                {
                    m_timerList.Add(delayObject as TimerItemBase);
                }
            }
        }

        protected override void removeObject(IDelayHandleItem delayObject)
        {
            // 检查当前是否在队列中
            if (m_timerList.IndexOf(delayObject as TimerItemBase) != -1)
            {
                (delayObject as TimerItemBase).m_disposed = true;
                if (bInDepth())
                {
                    base.removeObject(delayObject);
                }
                else
                {
                    foreach (TimerItemBase item in m_timerList.list())
                    {
                        if (UtilApi.isAddressEqual(item, delayObject))
                        {
                            m_timerList.Remove(item);
                            break;
                        }
                    }
                }
            }
        }

        // 从 Lua 中添加定时器，这种定时器尽量整个定时器周期只与 Lua 通信一次
        public void addTimer(TimerItemBase delayObject, float priority = 0.0f)
        {
            this.addObject(delayObject, priority);
        }

        public void addTimer(LuaTable luaTimer)
        {
            LuaTable table = luaTimer["pthis"] as LuaTable;
            LuaFunction function = luaTimer["func"] as LuaFunction;

            TimerItemBase timer = new TimerItemBase();
            timer.m_totalTime = Convert.ToSingle(luaTimer["totaltime"]);
            timer.setLuaFunctor(table, function);

            this.addTimer(timer);
        }

        public void removeTimer(TimerItemBase timer)
        {
            this.removeObject(timer);
        }

        public void Advance(float delta)
        {
            incDepth();

            foreach (TimerItemBase timerItem in m_timerList.list())
            {
                if (!timerItem.getClientDispose())
                {
                    timerItem.OnTimer(delta);
                }

                if (timerItem.m_disposed)        // 如果已经结束
                {
                    removeObject(timerItem);
                }
            }

            decDepth();
        }
    }
}